transformations/transformations/parallel_routine_dispatch.py

```diff
@@ -56,3 +99,146 @@ def create_dr_hook_calls(scope, cdname, pkey):
+        map_dcl = {}
+        for decl in routine.declarations:
+            if len(decl.symbols) == 1:
+                var = decl.symbols[0]
+                if var.name in self.routine_map_temp:
+                    new_vars = self.routine_map_temp[var.name]
+                    new_vars[1].type = new_vars[1].type.clone(pointer=True, shape=new_vars[1].dimensions, initial="NULL()")
+                    map_dcl.update({decl : (ir.VariableDeclaration(symbols=(new_vars[0],)), ir.VariableDeclaration(symbols=(new_vars[1],)))})
+            else:
+                raise Exception("Declaration should have only one symbol, please run single_variable_declaration before calling add_temp function.")
+        routine.spec = Transformer(map_dcl).visit(routine.spec)
+        #decls_to_replace = [var for var in decl.var for decl in routine.declarations if var in self.routine_map_temp]
+    
+    def add_field(self, routine):
+        # Insert the field generation wrapped into a DR_HOOK call
+        dr_hook_calls = self.create_dr_hook_calls(
+            routine, cdname='CREATE_TEMPORARIES',
+            handle=sym.Variable(name='ZHOOK_HANDLE_FIELD_API', scope=routine)
+        )
+        routine.body.insert(2, (dr_hook_calls[0], ir.Comment(text=''), *self.new_calls, dr_hook_calls[1]))
+
+        # Insert the field deletion wrapped into a DR_HOOK call
+        dr_hook_calls = self.create_dr_hook_calls(
+            routine, cdname='DELETE_TEMPORARIES',
+            handle=sym.Variable(name='ZHOOK_HANDLE_FIELD_API', scope=routine)
+        )
+        routine.body.insert(-2,(dr_hook_calls[0], ir.Comment(text=''), *self.delete_calls, dr_hook_calls[1]))
+
+    def decl_derived_types(self, routine, region):
+        region_map_derived = {}
+        derived = [var for var in FindVariables().visit(region) if var.name_parts[0] in routine.arguments]
+        for var in derived :
+            
+            key = f"{routine.variable_map[var.name_parts[0]].type.dtype.name}%{'%'.join(var.name_parts[1:])}"
+            if key in self.map_index:
+                value = self.map_index[key]
+                # Creating the pointer on the data : YL_A
+                data_name = f"Z_{var.name.replace('%', '_')}"
+                if "REAL" and "JPRB" in value[0]:
+                    data_type = SymbolAttributes(
+                        dtype=BasicType.REAL, kind=routine.symbol_map['JPRB'],
+                        pointer=True
+                    )
+                    data_dim = value[2] + 1
+                    data_shape = (sym.RangeIndex((None, None)),) * data_dim
+                    ptr_var = sym.Variable(name=data_name, type=data_type, dimensions=data_shape, scope=routine)
+
+                else:
+                    raise NotImplementedError("This type isn't implemented yet")
+
+                # Creating the pointer on the field api object : YL%FA, YL%F_A...
+                if routine.variable_map[var.name_parts[0]].type.dtype.name=="MF_PHYS_SURF_TYPE":
```

Comment:
Just as a remark: Using `routine.variable_map` repeatedly can be costly, because it compiles the map everytime from the variable declarations. In a routine like this, it's recommended to cache this at the top of the routine as `variable_map = routine.variable_map` and then use that local copy.